### PR TITLE
[BEAM-11032] Use metric for Java BigQuery streaming insert API latenc…

### PR DESCRIPTION
**Pull request by ihji**
…y logging

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ihji** [comment]
R: @ajamato

Uses Metrics for logging latencies instead of parameters. Reverted BigQueryService API signature changes for histogram and counter.

**Comment by ihji** [comment]
@ajamato PTAL

**Comment by ihji** [comment]
@ajamato Friendly ping

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/LatencyRecordingHttpRequestInitializer.java [hunk 1]
```diff
@@ -23,14 +23,26 @@
...
 import com.google.api.client.http.HttpResponseInterceptor;
 import java.io.IOException;
-import org.apache.beam.sdk.util.Histogram;
+import org.apache.beam.sdk.metrics.Histogram;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.util.HistogramData;
 
 /** HttpRequestInitializer for recording request to response latency of Http-based API calls. */
 public class LatencyRecordingHttpRequestInitializer implements HttpRequestInitializer {
+  public static final String HISTOGRAM_NAME = "latency_ms";
```
Comment (ajamato) [inline sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/LatencyRecordingHttpRequestInitializer.java:32]: Would you please use a specific name for the metric where it is instantiated

See the URN defined here:
API_REQUEST_LATENCIES
urn = beam:metric:io:api_request_latencies:v1
https://s.apache.org/beam-gcp-debuggability

For Labels:
for SERVICE_NAME use "BigQuery", METHOD_NAME you can use "BigQueryBatchWrite" and "BigQueryBatchRead". Which is just referring to API calls which read or write elements in batches.

Additional Labels (Populate these if they are available, if not leave it for follow up work for me):
BIGQUERY_PROJECT_ID
BIGQUERY_DATASET
BIGQUERY_TABLE
BIGQUERY_VIEW
BIGQUERY_QUERY_NAME- user provided query name
RESOURCE - combination of the above (See my python PR linked below for reference)

https://github.com/apache/beam/pull/13217/files#diff-5427a5d3887eb695cefde082c58575a2372972996b547d55961abeb4f7bc3debR576

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metrics.java [hunk 2]
```diff
@@ -96,6 +97,21 @@ public static Gauge gauge(Class<?> namespace, String name) {
     return new DelegatingGauge(MetricName.named(namespace, name));
   }
 
+  public static class Internal {
+    public static Counter counter(String namespace, String name, boolean processWideContainer) {
+      return new DelegatingCounter(MetricName.named(namespace, name), processWideContainer);
+    }
+
+    public static Histogram histogram(
+        String namespace,
```
Comment (ajamato) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metrics.java:106]: namespace and name parameters are specific to user metric (Which is a metric that populates a specific MonitoringInfo URN with a name and namspace parameter).

Have this just take a MetricName instead which allows for non user metrics
I.e. like the ElementCount metric
which is uniquely identifiers by a URN and labels.

https://github.com/apache/beam/pull/7272/files#diff-3834ef03b753005ee8251870a88fe3505732e9a7ad34da600f78e33ab9f37139R45

You can make it under the LabeledMetrics class, which is meant to be an internal implementation for SDK harness implementations. Not part of the public API (Much like the internal.Metrics in python you've added).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsLogger.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import org.slf4j.Logger;
+
+@Experimental(Kind.METRICS)
+public interface MetricsLogger extends Serializable {
+  Lock REPORTING_LOCK = new ReentrantLock();
+  AtomicLong LAST_REPORTED_MILLIS = new AtomicLong(System.currentTimeMillis());
+
+  default void tryLoggingMetrics(
+      String header, String namespace, long minimumLoggingFrequencyMillis, boolean resetMetrics) {
+    if (REPORTING_LOCK.tryLock()) {
```
Comment (ajamato) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsLogger.java:36]: Is it possible to use this style for logging, if it exists? If not ignore this suggestion.


        try (Autolock alock = new Autolock(lock)) {
            // Whatever you need to do while you own the lock
        } //unlocks here
        
        See:

https://stackoverflow.com/questions/151917/autolock-in-java-how-to

**Comment by ajamato** [comment]
@pabloem Would you take a look as well, as a committer. Or lmk if you have someone else in mind on the java side.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java [hunk 4]
```diff
@@ -38,7 +44,10 @@ public void inc() {
   /** Increment the counter by the given amount. */
   @Override
   public void inc(long n) {
-    MetricsContainer container = MetricsEnvironment.getCurrentContainer();
+    MetricsContainer container =
+        processWideContainer
```
[the remark this answers is not in the document]
Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metrics.java (hunk 2 above)
Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metrics.java]: Done. Moved into `LabeledMetrics`.

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/LatencyRecordingHttpRequestInitializer.java (hunk 1 above)
Comment (ihji) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/util/LatencyRecordingHttpRequestInitializer.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsLogger.java (hunk 3 above)
Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsLogger.java:40]: `Lock` is not auto closeable by default and IMHO `tryLock` doesn't fit well with try-with-resources style (it hides the semantics of `tryLock`).

**Comment by ihji** [comment]
Run Java PreCommit

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java [hunk 5]
```diff
@@ -383,4 +409,70 @@ public boolean equals(@Nullable Object object) {
   public int hashCode() {
     return Objects.hash(stepName, counters, distributions, gauges);
   }
+
+  private boolean equalsMetricName(MetricName metricName, String namespace, String name) {
+    return (namespace == null || namespace.equals(metricName.getNamespace()))
```
Comment (ajamato) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java:414]: This equality looks werid, why would it be considered equal if the param is null?

Can you add a clairifying comment

**Review by ajamato — APPROVED** [review]

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java (hunk 5 above)
Comment (ihji) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java:419]: Yeah, `equals` prefix is misleading. It's more like matching and when the field will be ignored when the parameter is null. I renamed the method.

**Comment by ihji** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by ihji** [comment]
Run Java PreCommit

**Comment by ihji** [comment]
Run Java PreCommit

**Pull request merged**